mitmproxy/proxy/layers/http/__init__.py

```diff
@@ -756,6 +756,15 @@ def handle_connect_finish(self):
                 self.context.client,
             )
             yield SendHttp(ResponseEndOfMessage(self.stream_id), self.context.client)
+        elif self.flow.response.status_code == 502:
+            self.flow.error = flow.Error("CONNECT failed")
+            yield HttpErrorHook(self.flow)
```

Comment:
I realize the original issue wasn't that clear here: I think we neither want to call `response` nor `error` here - these are reserved for regular flows. We don't want flows to magically appear here without previously going through `requestheaders` first. Instead, we should add an `http_connect_error` hook (see comment in the test below).